### DISCUSSION TITLE
Merkel Tree / branching diagram

### DISCUSSION THREAD

**Discussion by gruckion** [open-webui/open-webui#1643] [Ideas]
# Branching conversation visual aid

I propose we implement a simple branching diagram similar to [Grok](https://en.rattibha.com/thread/1721053802235621734). This UI approach will make it easier to have branching conversations.

<img width="2052" alt="Screenshot 2024-04-21 at 10 50 31" src="https://github.com/open-webui/open-webui/assets/4559650/f4e996d2-c9fb-4fcf-9334-d384d7262d7c">

https://github.com/open-webui/open-webui/assets/4559650/19fa8039-a7f5-4dc3-b9f2-3286cac87f0b
Credit: [@TobyPhln](https://twitter.com/TobyPhln)

> E.g. you've shown the LLM your code and now you want to generate unit tests for each of say 12 functions. But doing this linearly will result in the context window shifting. A better approach is to edit your previous message but this means you are loosing your down stream changes each time you click edit.

<details>
  <summary>@discord/garlicbread726 posted the following.</summary>
  
# Garlicbread726 project enhancements

I would to share some ideas this project can enhance it's interface in upcoming versions:

1. Introduce a file management system for effortless navigation and retrieval of chat logs.
2. Implement a Merkel Tree view for chats, enabling swift identification of chat regenerations.
3. Launch a Round Table Mode, simplifying the process of group chatting with agents.

For a clearer understanding, I've prepared some Figma sketches...

## File Management

Introducing a file management system within the interface could significantly streamline the process of navigating and retrieving chat logs. Currently, there is a noticeable amount of unused space in the sidebar, which presents a perfect opportunity to incorporate a folder tree. This addition would allow users to organize chat logs in a structured manner, making it easier to locate specific conversations or topics without the need to sift through a long list of unsorted chat entries. By leveraging this unused sidebar space for a folder tree, users can categorize their chats into folders and subfolders, enabling quick access to any chat log with just a few clicks. This improvement not only enhances user experience by simplifying the retrieval process but also makes the interface more efficient and user-friendly, catering to the needs of users who manage a large volume of chats.

![Add_Folders](https://github.com/open-webui/open-webui/assets/4559650/c3da5245-68da-438e-ab9e-a2af842c9956)

## Merkel Tree

Implementing a Merkel Tree view for chats represents a transformative step towards enabling users to swiftly identify and navigate through chat regenerations. The current user interface presents a significant challenge when it comes to locating forks in chat conversations. Users often find it frustrating to trace the evolution of discussions, especially when conversations diverge into multiple threads or regenerations. This difficulty can impede the brainstorming process, where the ability to quickly visualize and access different branches of a conversation is crucial.

A Merkel Tree, by its design, organizes data in a hierarchical structure, making it an ideal solution for visualizing the lineage of chat regenerations. By adopting this approach, each chat and its subsequent forks can be represented as nodes within the tree. This allows users to see at a glance how conversations have evolved, identify specific points where forks occurred, and directly access these regenerations without having to navigate through a cumbersome linear history.

Incorporating a Merkel Tree view into the chat interface would significantly enhance the user experience by providing a clear, organized way to track the progression of discussions. This improvement would be particularly valuable for users engaged in creative or collaborative processes, where the ability to explore different conversational pathways and revisit ideas is key to developing innovative solutions.

![Merkle_Tree_1](https://github.com/open-webui/open-webui/assets/4559650/19aa49aa-b2e6-48de-920d-2a092b571f18)

![Merkle_Tree_2](https://github.com/open-webui/open-webui/assets/4559650/84f29e8e-5707-4b0d-8edd-aa088f6984b1)

## Round Table

Launching a Round Table Mode would revolutionize the way users interact with multiple agents simultaneously, addressing a notable frustration with the current user interface: the limitation of engaging with agents linearly. This restriction can be particularly cumbersome for users who seek to leverage the diverse capabilities and perspectives of multiple agents for more complex discussions or collaborative brainstorming sessions.

The proposed Round Table Mode aims to simplify the process of group chatting with agents by introducing a more dynamic and interactive framework. Within this mode, users would have the flexibility to select a group of agents to participate in a single conversation. This inclusive approach would facilitate a richer, more multifaceted dialogue, mimicking a real-world round table discussion.

Key features of the Round Table Mode would include the ability to turn on and off auto-responses for individual agents, providing users with control over the flow of the conversation. Moreover, users could decide whether responses from agents are added directly to the round table for immediate consideration or moved to a queue for later review. This level of customization ensures that the conversation remains organized and focused, aligning with the user's specific needs and preferences.

Implementing such a feature would not only enhance the user experience by removing existing barriers to multi-agent interaction but also position the product as a unique and innovative platform for collaborative discussion. By fostering a more engaging and versatile chat environment, Round Table Mode has the potential to attract a wider user base and cultivate a vibrant community around the product, encouraging users to explore new ways of generating ideas and solving problems together.

![Round_Table](https://github.com/open-webui/open-webui/assets/4559650/7518084a-3b31-424c-a593-01e9dcc9dcc8)

[Discord link](https://discord.com/channels/1170866489302188073/1171573342956965919/1213343361520640071). Side note, the visual are very nice and well put together, nice work garlicbread726.
</details>

## Technical requirements

in progress

**Comment by thiswillbeyourgithub** [comment]
Btw here's the neat docs of librechat about forking:
https://www.librechat.ai/docs/user_guides/fork
